staging/src/k8s.io/kube-aggregator/pkg/controllers/status/metrics.go

```diff
@@ -38,17 +44,65 @@ var (
 		},
 		[]string{"name", "reason"},
 	)
-	unavailableGauge = metrics.NewGaugeVec(
-		&metrics.GaugeOpts{
-			Name:           "aggregator_unavailable_apiservice",
-			Help:           "Gauge of APIServices which are marked as unavailable broken down by APIService name.",
-			StabilityLevel: metrics.ALPHA,
-		},
+
+	unavailableGaugeDesc = metrics.NewDesc(
+		"aggregator_unavailable_apiservice",
+		"Gauge of APIServices which are marked as unavailable broken down by APIService name.",
 		[]string{"name"},
+		nil,
+		metrics.ALPHA,
+		"",
 	)
 )
+var registerMetrics sync.Once
+
+// Register registers metrics for status controller.
+func Register(apiServiceLister listers.APIServiceLister) {
+	registerMetrics.Do(func() {
+		legacyregistry.CustomMustRegister(newAPIServiceStatusCollector(apiServiceLister))
+		legacyregistry.MustRegister(unavailableCounter)
+	})
+}
+
+type apiServiceStatusCollector struct {
+	metrics.BaseStableCollector
+
+	apiServiceLister listers.APIServiceLister
+}
+
+// Check if apiServiceStatusCollector implements necessary interface.
+var _ metrics.StableCollector = &apiServiceStatusCollector{}
+
+func newAPIServiceStatusCollector(apiServiceLister listers.APIServiceLister) *apiServiceStatusCollector {
+	return &apiServiceStatusCollector{
+		apiServiceLister: apiServiceLister,
+	}
+}
+
+// DescribeWithStability implements the metrics.StableCollector interface.
+func (c *apiServiceStatusCollector) DescribeWithStability(ch chan<- *metrics.Desc) {
+	ch <- unavailableGaugeDesc
+}
 
-func init() {
-	legacyregistry.MustRegister(unavailableCounter)
-	legacyregistry.MustRegister(unavailableGauge)
+// CollectWithStability implements the metrics.StableCollector interface.
+func (c *apiServiceStatusCollector) CollectWithStability(ch chan<- metrics.Metric) {
```

Comment:
when is this called?

Reply:
This is called when the apiserver metrics are scraped. The exact interval between each scrape depends on how it is configured, but it's usually every 30s.

Reply:
ack. It's what I expected.